pkg/block/block.go

```diff
@@ -99,6 +99,12 @@ func Download(ctx context.Context, logger log.Logger, bucket objstore.Bucket, id
 // TODO(bplotka): Ensure bucket operations have reasonable backoff retries.
 // NOTE: Upload updates `meta.Thanos.File` section.
 func Upload(ctx context.Context, logger log.Logger, bkt objstore.Bucket, bdir string, hf metadata.HashFunc) error {
+	return UploadWithRetry(ctx, logger, bkt, bdir, hf, 0)
+}
+
+// UploadWithRetry is a utility function for upload and acts as a workaround for absence of default parameters (which in this case is retryCounter = 0).
+func UploadWithRetry(ctx context.Context, logger log.Logger, bkt objstore.Bucket, bdir string, hf metadata.HashFunc, retryCounter int) error {
```

Comment:
From the name `UploadWithRetry` I will think it runs the `Upload` function with retries, but that is not the case here. It would have been better to have an function `Upload(...)` which handles the upload logic without retry, then a `UploadWithRetries` function that retries this `Upload` fn itself.


But we are only trying to retry when upload fails, so instead of having retries here, doesn't it makes more sense to have a `objstore.UploadDirWithRetries` that will handle this transparently.

Reply:
It is doing exactly that. Isn't it ? @prmsrswt 😕

Reply:
The problem with a separate function would be : we'd have to carry over the error which caused a failure, and keeping track of which error caused it to fail would become difficult otherwise. The only reason I made the UploadWithRetries function was to keep the integrity of Upload intact and not add any other parameter to it. 
It would have been really great if Golang supported default params 😁